### PR TITLE
DD-30: Creation of Direct Debit Message Templates

### PR DESCRIPTION
**Pull request by awg-github**
## Overview

1. Add the following user driven message templates when install the extension:
- Direct Debit Payment Sign Up Notification
- Direct Debit Payment Update Notification
- Direct Debit Payment Collection Reminder
- Direct Debit Auto-renew Notification
- Direct Debit Mandate Update Notification

2. All message templates contain:
- the comments and text shown in the screenshot
- some smarty code for generating dynamic informations
- some HTML code for formatting the letter and direct debit guarantee

3. For simplicity, all five templates include all the following dynamic information:
- Recurring contribution information
- Direct Debit Mandate (with condition to hide if payment method is not Direct Debit)
- Membership Information
- Next contribution information
- Direct Debit Guarantee (with condition to hide if payment method is not Direct Debit)

dd-30 direct debit message templates1

### CODE REVIEW

File: CRM/ManualDirectDebit/Mail/Task/Mail.php [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+      )
+      ) {
+        $sent = TRUE;
+      }
+    }
+
+    return array($sent, $activity->id);
+  }
+
+}
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Mail/Task/Mail.php]: Maybe jst add a blank line at the end of this file.

File: CRM/ManualDirectDebit/Mail/Task/MembershipEmailCommon.php [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+        $validatedIds[] = (int) $id;
+      }
+
+      return $validatedIds;
+    }
+
+    return [];
+  }
+
+}
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Mail/Task/MembershipEmailCommon.php]: Missing blank line at EOF.

File: templates/CRM/ManualDirectDebit/MessageTemplate/PaymentUpdateNotification.tpl [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                    </ul>
+                </td>
+                <td></td>
+            </tr>
+        </table>
+        <!-- End of Direct Debit Guarantee.  -->
+    {/if}
+</div>
+</body>
+</html>
```
Comment (MiyaNoctem) [inline templates/CRM/ManualDirectDebit/MessageTemplate/PaymentUpdateNotification.tpl]: Missing blank line

File: templates/CRM/ManualDirectDebit/MessageTemplate/PaymentSignUpNotification.tpl [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+                    </ul>
+                </td>
+                <td></td>
+            </tr>
+        </table>
+        <!-- End of Direct Debit Guarantee.  -->
+    {/if}
+</div>
+</body>
+</html>
```
Comment (MiyaNoctem) [inline templates/CRM/ManualDirectDebit/MessageTemplate/PaymentSignUpNotification.tpl]: Missing blank line at EOF.

File: templates/CRM/ManualDirectDebit/MessageTemplate/PaymentCollectionReminder.tpl [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+                    </ul>
+                </td>
+                <td></td>
+            </tr>
+        </table>
+        <!-- End of Direct Debit Guarantee.  -->
+    {/if}
+</div>
+</body>
+</html>
```
Comment (MiyaNoctem) [inline templates/CRM/ManualDirectDebit/MessageTemplate/PaymentCollectionReminder.tpl]: Missing blank line at EOF.

File: templates/CRM/ManualDirectDebit/MessageTemplate/MandateUpdateNotification.tpl [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+                    </ul>
+                </td>
+                <td></td>
+            </tr>
+        </table>
+        <!-- End of Direct Debit Guarantee.  -->
+    {/if}
+</div>
+</body>
+</html>
```
Comment (MiyaNoctem) [inline templates/CRM/ManualDirectDebit/MessageTemplate/MandateUpdateNotification.tpl]: Missing blank line at EOF.

File: templates/CRM/ManualDirectDebit/MessageTemplate/AutoRenewNotification.tpl [hunk 7]
```diff
@@ -0,0 +1,163 @@
...
+                    </ul>
+                </td>
+                <td></td>
+            </tr>
+        </table>
+        <!-- End of Direct Debit Guarantee.  -->
+    {/if}
+</div>
+</body>
+</html>
```
Comment (MiyaNoctem) [inline templates/CRM/ManualDirectDebit/MessageTemplate/AutoRenewNotification.tpl]: Missing blank line at EOF.

File: templates/CRM/ManualDirectDebit/Form/Email/Membership.tpl [hunk 8]
```diff
@@ -0,0 +1 @@
+{include file="CRM/Contact/Form/Task/Email.tpl"}
```
Comment (MiyaNoctem) [inline templates/CRM/ManualDirectDebit/Form/Email/Membership.tpl]: Missing blank line at EOF.

File: CRM/ManualDirectDebit/Common/MessageTemplate.php [hunk 9]
```diff
@@ -0,0 +1,68 @@
+<?php
+
+/**
+ * Checks if template is direct debit template
+ */
+class CRM_ManualDirectDebit_Common_MessageTemplate {
+
+  const SING_UP = 'Direct Debit Payment Sign Up Notification';
```
[the remark this answers is not in the document]
Comment (awg-github) [reply CRM/ManualDirectDebit/Common/MessageTemplate.php]: Sorry, can you provide more information for this comment

File: CRM/ManualDirectDebit/Mail/Notification.php [hunk 10]
```diff
@@ -73,19 +73,8 @@ private function setSenderEmailAddress($senderEmailAddress) {
...
-        CRM_ManualDirectDebit_Common_MessageTemplate::SING_UP
-      );
-
-      return $this->sendEmail($email, $messageTemplateId, $tplParams);
-    }
-    catch (CiviCRM_API3_Exception $e) {
-      return FALSE;
-    }
+    $dataCollector = new CRM_ManualDirectDebit_Mail_DataCollector_RecurringContribution($recurringContributionId);
+    return $this->sendEmail($dataCollector, CRM_ManualDirectDebit_Common_MessageTemplate::SIGN_UP_MSG_TITLE);
```
Comment (MiyaNoctem) [inline CRM/ManualDirectDebit/Mail/Notification.php]: Please add a blank line before the return statement. Also for the rest of notify functions.

**Pull request merged**